### PR TITLE
workflows: Trigger anaconda test on bridge/storaged changes

### PR DESCRIPTION
**Pull request by martinpitt**
We want to make sure to not break Anaconda with changes which affect it, i.e. the bridge or the Storage page. As Anaconda's tests are "special" (require booting boot.iso, can't run in tmt/Testing Farm), we need to run them in Cockpit's CI.

Add a workflow which runs if the PR affects Anaconda (changes to the bridge or Storage page), polls the packit COPR until it has the current PR version available, and then test-triggers a "cockpit PR" scenario.

https://issues.redhat.com/browse/COCKPIT-1064

----

This got tested in #19459 with a demo commit that touches src/bridge, and got anaconda triggered. In *this* PR, anaconda should not be triggered, and the workflow succeed quickly with determining "not affected".

### CODE REVIEW

**Review by jelly — APPROVED** [review]
It feels to me that we could test anaconda in packit, but that is a different topic

**Comment by martinpitt** [comment]
@jelly Yeah, long story.. I don't think it'll happen anytime soon, and I'd like to have something working now.

@allisonkarlitskaya I know you have strong opinions about workflows, so I'll keep this open and wait for your review as well.

**Review by KKoukiou — COMMENTED** [review]
The changes look good. Please help me understand the new testing integration workflow though a bit.

Facts:
1. Integration tests between cockpit storage and anaconda would go to the anaconda repo
2. When the storage page changes in an expected way and the tests get broken.

Qs:
1.What's the process of updating these? I guess we send a PR to anaconda repo to update, but since we test against the cockpit from rawhide by default, we can't merge that PR till you make a release.

**Comment by martinpitt** [comment]
@KKoukiou We have to figure this out still, as none of that integration exists on our main branches yet. My understanding is that:

 F1. is *mostly* not what was intended -- I understand that the "anaconda mode" (chroot prefix, different UI) tests would stay in the cockpit repo and co-evolve with the code, and anaconda just tests the interface (calling the storage page, and ensuring it gets a proper JSON reply about the final layout), so that..
 F2. in most cases changes to the Storage page shouldn't actually break anaconda.

Q1. However: our CI has the power of running a non-main branch in a "foreign" project. We do this all the time with e.g. testing a cockpit branch against a bots branch for updating an image refresh with a lockstep code change. Of course these can't/won't be triggered automatically, but I see the process as :

   1. file cockpit PR that changes the storage page; anaconda/main test fails
   2. file an Anaconda PR to do the corresponding code/test adjustments
   3. trigger a test against that anaconda PR in the cockpit PR, which should go green. This will require inventing some `tests-trigger` syntax, but all the technical requirements are already there.

WDYT?

**Comment by KKoukiou** [comment]
@martinpitt sounds good.

In order to prepare all possible storage scenarios, we would need to execute test code against the cockpit-storage from the anaconda tests. For keeping the anaconda tests a bit detaches from the details of the cockpit-storage, cockpit could export in testlib some 'storage helper functions' like, 'def create_new_partition', 'def format_device' etc, so that we can reuse the high level methods without caring about the selectors of the cockpit-storage.

We will see this in practice.

File: .github/workflows/trigger-anaconda.yml
Comment (allisonkarlitskaya) [inline .github/workflows/trigger-anaconda.yml:39]: Sounds like it's better accomplished by way of `on.pull_request_target.paths`: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore

File: .github/workflows/trigger-anaconda.yml
Comment (allisonkarlitskaya) [inline .github/workflows/trigger-anaconda.yml:63]: Surprised we don't see something like `@cockpit#12345` to match what we do with the `@bots#` syntax.  We could teach `tests-scan` about that, or just leave it alone and let Anaconda handle it for itself...

**Review by allisonkarlitskaya — COMMENTED** [review]
I sort of dislike the whole approach to be honest.  No doubt it will work, and no doubt that it's the least-effort way to get what we need right now, but we've talked about queuing statuses from workflows in the path and have always shied away from the idea...

I guess my preferred solution to this problem would be making the webhook more intelligent and adjusting the testmap to allow for path-based cross-testing, but that's a way bigger change....

**Comment by allisonkarlitskaya** [comment]
Note also, this might be useful: https://github.com/cockpit-project/cockpit/pull/19460.diff

You can fetch the diff and easily peel the list of affected files out of it.

**Comment by martinpitt** [comment]
> making the webhook more intelligent and adjusting the testmap to allow for path-based cross-testing

FTR: The test needs to start once the COPR build is done, not when the PR gets pushed. Of course our test could do the polling, but I wanted to avoid blocking one of our bots with just twiddling thumbs for 5 to 30 minutes, and offload this to GitHub. I can't say I like the token thing much, but I don't dislike it enough to sacrifice bots capacity.

File: .github/workflows/trigger-anaconda.yml
Comment (martinpitt) [reply .github/workflows/trigger-anaconda.yml]: Thanks for pointing out! This is so much nicer indeed, then the workflow doesn't even start nor show up in the checks list for PRs which don't affect anaconda. I updated the workflow accordingly, and tested both ways in #19459 again.

File: .github/workflows/trigger-anaconda.yml
Comment (martinpitt) [reply .github/workflows/trigger-anaconda.yml]: This wouldn't quite fit, as we already trigger a foreign project with `@rhinstaller/anaconda`. We have to let this run in a cockpit PR: There is no anconda PR to run it in, we don't want to attach the status to anaconda/master (as that would collide with multiple cockpit PRs and interfere with anaconda nightly test runs), and most importantly, we actually want to see the anaconda status in the cockpit PR.

In theory we could have a syntax like `fedora-rawhide-boot@rhinstaller/anaconda@cockpit#12345`, or `fedora-rawhide-boot@cockpit#12345@rhinstaller/anaconda`, but IMHO this is both confusing and hard to parse. So for now I'd just go with the syntax that we have and avoid blocking this on bots changes.

**Comment by martinpitt** [comment]
> we've talked about queuing statuses from workflows in the path and have always shied away from the idea...

(I assume s/path/past/). In the past I had some PR which triggered tests from the dist/ cache (jumpstart) workflow: i.e. it would let that finish, and then run tests-scan on the PR with `--amqp-server`. That would have entirely short-circuited the webhook, and thus dropping several moving parts, in exchange for having to put the AMQP queuing certificate into the github secrets. I don't remember any more why we didn't do that in the end, but I suspect it was because we still need the webhook anyway for developers running `bots/tests-trigger`; also, these days we don't have jumpstart any more anyway.

Do you still remember why we "shied away from the idea"? I don't. Thanks!

File: .github/workflows/trigger-anaconda.yml
Comment (allisonkarlitskaya) [inline .github/workflows/trigger-anaconda.yml:53]: I still think we could be more clever here....

For example, when we run `...@cockpit-project/other` PR runs from bots PRs the tests infra is smart enough to understand that we want to test that bots PR...

**Review by allisonkarlitskaya — APPROVED** [review]
I think maybe the idea we discussed before was rather to eliminate the webhook and do everything via direct queuing from workflows...

I still think it would be nicer to evaluate this from `tests-scan`, but I don't want to block this PR any longer.

A thought: do we get webhook events when packit posts the result from the RPM build completion?  Could we?  Maybe we could check at that point if we want to trigger anything...

File: .github/workflows/trigger-anaconda.yml
Comment (martinpitt) [reply .github/workflows/trigger-anaconda.yml:53]: I'm not sure what you mean here? That already happens (it sets `$COCKPIT_BOTS_REF`). The `@bots#` syntax is for the other direction: Run a test in a *cockpit* PR against a bots PR instead of bots main.

**Comment by martinpitt** [comment]
> I think maybe the idea we discussed before was rather to eliminate the webhook and do everything via direct queuing from workflows...

Yes, that's what I remember (see above).

>  Do we get webhook events when packit posts the result from the RPM build completion? Could we? Maybe we could check at that point if we want to trigger anything...

That's a nice idea. More bots work to do for this special case and more moving parts, but it would avoid the ugly polling and the cockpituous token. I captured that in https://issues.redhat.com/browse/COCKPIT-1071.

I'd land this now to unblock the anaconda work for Marius, as implementing the above will take some time. But that epic is in my quarterly plan, so I won't forget it accidentally.

**Pull request merged**
